[Profiler] Defer recording startup python events (take 2)

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91684

This is my commandeer of https://github.com/pytorch/pytorch/pull/82154 with a couple extra fixes.

The high level idea is that when we start profiling we see python frames which are currently executing, but we don't know what system TID created them. So instead we defer the TID assignment, and then during post processing we peer into the future and use the system TID *of the next* call on that Python TID.

As an aside, it turns out that CPython does some bookkeeping (https://github.com/python/cpython/blob/ee821dcd3961efc47262322848267fe398faa4e4/Include/cpython/pystate.h#L159-L165, thanks @dzhulgakov for the pointer), but you'd have to do some extra work at runtime to know how to map their TID to ours so for now I'm going to stick to what I can glean from post processing alone.

As we start observing more threads it becomes more important to be principled about how we start up and shut down. (Since threads may die while the profiler is running.) #82154 had various troubles with segfaults that wound up being related to accessing Python thread pointers which were no longer alive. I've tweaked the startup and shutdown interaction with the CPython interpreter and it should be safer now.

Differential Revision: [D42336292](https://our.internmc.facebook.com/intern/diff/D42336292/)

Commit:
[Profiler] Defer recording startup python events (take 2)

This is my commandeer of https://github.com/pytorch/pytorch/pull/82154 with a couple extra fixes.

The high level idea is that when we start profiling we see python frames which are currently executing, but we don't know what system TID created them. So instead we defer the TID assignment, and then during post processing we peer into the future and use the system TID *of the next* call on that Python TID.

As an aside, it turns out that CPython does some bookkeeping (https://github.com/python/cpython/blob/ee821dcd3961efc47262322848267fe398faa4e4/Include/cpython/pystate.h#L159-L165, thanks @dzhulgakov for the pointer), but you'd have to do some extra work at runtime to know how to map their TID to ours so for now I'm going to stick to what I can glean from post processing alone.

As we start observing more threads it becomes more important to be principled about how we start up and shut down. (Since threads may die while the profiler is running.) #82154 had various troubles with segfaults that wound up being related to accessing Python thread pointers which were no longer alive. I've tweaked the startup and shutdown interaction with the CPython interpreter and it should be safer now.

Differential Revision: [D42336292](https://our.internmc.facebook.com/intern/diff/D42336292/)

[ghstack-poisoned]

albanD commented on test/profiler/test_profiler.py:
```diff
@@ -545,6 +549,161 @@ def extract(pattern: str):
+                (True, True),
+
+            ),
+        }.items(),
```
Ok!
The change looks good from python API side. I'll let someone else review the profiler logic!

robieta replied:
Thanks!

dzhulgakov commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -938,6 +975,20 @@ class PostProcess {
+    auto it = out.rbegin();
+    for (C10_UNUSED auto _ : c10::irange(initial_size, out.size())) {
+      const auto python_tid =
+          c10::get<ExtraFields<E>>((*it)->extra_fields_).python_tid_;
```
if you want to be really paranoid, Python thread ids are just results of `pthread_self` on linux: https://github.com/python/cpython/blob/2f2fa03ff3d566b675020787e23de8fb4ca78e99/Python/thread_pthread.h#L313 (and native_thread_id was added in 3.11 only). So they technically could be reused between invocations.

If you did bookkeeping on RecordFunction::currentThreadId <-> pthread_self level I guess it'd be safer (as you'd do lookups only while the thread is active), but it might be an overkill.

dzhulgakov commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -938,6 +975,20 @@ class PostProcess {
+    // observed event with the same Python TID.
+    ska::flat_hash_map<size_t, size_t> tid_map;
+    auto it = out.rbegin();
+    for (C10_UNUSED auto _ : c10::irange(initial_size, out.size())) {
```
nit: why not a good old iteration from rbegin to rend()-initial_size?

dzhulgakov reviewed (approved):
My comments are unblocking, so accepting it, nice fix!

dzhulgakov commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -868,6 +892,18 @@ class PostProcess {
+      enters.push_back(
+          {frame.trace_key_,
+           NoTID, // Allows us to detect unhandled start frames
+           enters.begin()->kineto_info_,
```
While testing this on a pretty big job the crashes appear consistently on this line. `enters` is empty, though I can't repro it in small experiments.

In any case, this logic appears flawed. `kineto_info_` contains `thread_id` so it can't be set equal for events. We should do the same patching logic as for the thread id above to populate it. It seems that there's `Result::kineto_info_` preserved intact.

Side question: isn't the thread id in `kineto_info_` and at the top level the same? Why do we need both?

cc @jamesr66a

dzhulgakov replied:
Following up on this bug (independent of the GIL), it can happen that `enters` is empty, and then it'd crash in post-processing. Do you agree with my reasoning above?

robieta replied:
Yeah, this seems wrong. Both because of the crash and because it's logically incorrect. I have no idea how that got there. (Maybe it was meant to be a placeholder too?)

The crash happens if there are enter events, but no python executes in the profiled region. This is impossible with the context manager since it has to run `__exit__` at least, but not so for on-demand. (A simple repro is just to trace `sleep`.) So your workload must be doing something else (long C++ kernel, blocked waiting on dataloader, etc.) I'm just going to make it default construct for now since I don't think it's particularly load bearing. (And we can revisit if or when it matters.)

As for why TID is stored multiple places: it's possible to build profiler without Kineto. That means that Profiler can't use any Kineto structs in its core logic because if it does the `#ifdef`s will blow up the non-kineto build. (We always include one kineto header to get an enum, but even that was very painful.) So until there is any consolidation profiler handles its own logic and kineto metadata is mostly along for the ride.

Commit:
Update on "[Profiler] Defer recording startup python events (take 2)"


This is my commandeer of https://github.com/pytorch/pytorch/pull/82154 with a couple extra fixes.

The high level idea is that when we start profiling we see python frames which are currently executing, but we don't know what system TID created them. So instead we defer the TID assignment, and then during post processing we peer into the future and use the system TID *of the next* call on that Python TID.

As an aside, it turns out that CPython does some bookkeeping (https://github.com/python/cpython/blob/ee821dcd3961efc47262322848267fe398faa4e4/Include/cpython/pystate.h#L159-L165, thanks dzhulgakov for the pointer), but you'd have to do some extra work at runtime to know how to map their TID to ours so for now I'm going to stick to what I can glean from post processing alone.

As we start observing more threads it becomes more important to be principled about how we start up and shut down. (Since threads may die while the profiler is running.) #82154 had various troubles with segfaults that wound up being related to accessing Python thread pointers which were no longer alive. I've tweaked the startup and shutdown interaction with the CPython interpreter and it should be safer now.

Differential Revision: [D42336292](https://our.internmc.facebook.com/intern/diff/D42336292/)

[ghstack-poisoned]

dzhulgakov commented:
I got another crash testing this PR:

```
Fatal Python error: _PyThreadState_DeleteCurrent: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
Python runtime state: initialized

Current thread 0x00007f1bd26b5700 (most recent call first):
<no Python frame>

Thread 0x00007f1bd1db1700 (most recent call first):
<no Python frame>

Thread 0x00007f1c703b8180 (most recent call first):
  File "/home/dzhulgakov/work/pytorch/torch/autograd/__init__.py", line 197 in backward
  File "/home/dzhulgakov/work/pytorch/torch/_tensor.py", line 488 in backward
  File "/home/dzhulgakov/work/pytorch/../traces/trigger_crash.py", line 19 in <module>
Aborted (core dumped)
```

The stack coming out of python tracer:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f1c703db859 in __GI_abort () at abort.c:79
#2  0x000000000049ff80 in fatal_error_exit (status=<error reading variable: dwarf2_find_location_expression: Corrupted DWARF expression.>) at /usr/local/src/conda/python-3.9.13/Python/pylifecycle.c:2201
#3  fatal_error (stream=<optimized out>, header=header@entry=1, prefix=prefix@entry=0x68ba90 <__func__.6.lto_priv.4> "_PyThreadState_DeleteCurrent",
    msg=msg@entry=0x6260c8 "the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)", status=status@entry=-1)
    at /usr/local/src/conda/python-3.9.13/Python/pylifecycle.c:2219
#4  0x00000000004a02ea in _Py_FatalErrorFunc (func=func@entry=0x68ba90 <__func__.6.lto_priv.4> "_PyThreadState_DeleteCurrent",
    msg=msg@entry=0x6260c8 "the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)") at /usr/local/src/conda/python-3.9.13/Python/pylifecycle.c:2301
#5  0x00000000004a0325 in _Py_FatalError_TstateNULL (func=func@entry=0x68ba90 <__func__.6.lto_priv.4> "_PyThreadState_DeleteCurrent") at /usr/local/src/conda/python-3.9.13/Python/ceval.c:251
#6  0x000000000047456c in _PyThreadState_DeleteCurrent (tstate=<optimized out>) at /usr/local/src/conda/python-3.9.13/Include/internal/pycore_pystate.h:83
#7  0x00007f1c216e5fb5 in pybind11::gil_scoped_acquire::dec_ref() () from /home/dzhulgakov/work/pytorch/torch/lib/libtorch_python.so
#8  0x00007f1c216e5fcd in pybind11::gil_scoped_acquire::~gil_scoped_acquire() () from /home/dzhulgakov/work/pytorch/torch/lib/libtorch_python.so
#9  0x00007f1c21a5e7da in torch::profiler::impl::(anonymous namespace)::PythonTracer::stop() () from /home/dzhulgakov/work/pytorch/torch/lib/libtorch_python.so
#10 0x00007f1c1a938ebf in torch::autograd::profiler::(anonymous namespace)::KinetoThreadLocalState::finalizeTrace() () from /home/dzhulgakov/work/pytorch/torch/lib/libtorch_cpu.so
#11 0x00007f1c1a93b125 in torch::autograd::profiler::disableProfiler() () from /home/dzhulgakov/work/pytorch/torch/lib/libtorch_cpu.so
#12 0x00007f1c1b193791 in torch::profiler::impl::(anonymous namespace)::LibKinetoClient::stop() () from /home/dzhulgakov/work/pytorch/torch/lib/libtorch_cpu.so
#13 0x00007f1c1cb4305b in libkineto::CuptiActivityProfiler::performRunLoopStep(std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > > const&, std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > > const&, long) () from /home/dzhulgakov/work/pytorch/torch/lib/libtorch_cpu.so
#14 0x00007f1c1cb4c6ca in libkineto::ActivityProfilerController::profilerLoop() () from /home/dzhulgakov/work/pytorch/torch/lib/libtorch_cpu.so
#15 0x00007f1c26f75bf4 in std::execute_native_thread_routine (__p=0x7f1bdc007650) at /opt/conda/conda-bld/gcc-compiler_1654084175708/work/gcc/libstdc++-v3/src/c++11/thread.cc:82
#16 0x00007f1c7070d609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#17 0x00007f1c704d8133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

It reproduces sporadically even on a simplest test

```
import torch
import torchvision.models as models


model = models.resnet50().cuda()
loss = torch.nn.CrossEntropyLoss()

BS = 1
x = torch.randn(BS, 3, 224, 224, device='cuda').detach()
label = torch.empty(BS, 1000, device='cuda').detach()

for _ in range(100000):
    out = model(x)
    loss_val = loss(out, label)
    loss_val.backward()
```

And trigger it periodically with something like `while true; do dyno gputrace --log-file=/tmp/foo.pt.trace.json; sleep 7; done`. Running for a few mins is enough to trigger the crash

So I think something is still messed up with refcounting.

chaekit commented:
> while true; do dyno gputrace --log-file=/tmp/foo.pt.trace.json; sleep 7; done

QQ - is there a reason you are using on demand tracing via dyno rather than using Profiler API? we are trying to move away from time-based tracing in favor of step based tracing which gives clearer(often more stable) tracing experience

dzhulgakov commented:
It's mostly to avoid instrumenting the model code. Why are you trying to move away from time-based tracing? I find it's very convenient to see "what's going on" without having to change the code in any way.

Commit:
Update on "[Profiler] Defer recording startup python events (take 2)"


This is my commandeer of https://github.com/pytorch/pytorch/pull/82154 with a couple extra fixes.

The high level idea is that when we start profiling we see python frames which are currently executing, but we don't know what system TID created them. So instead we defer the TID assignment, and then during post processing we peer into the future and use the system TID *of the next* call on that Python TID.

As an aside, it turns out that CPython does some bookkeeping (https://github.com/python/cpython/blob/ee821dcd3961efc47262322848267fe398faa4e4/Include/cpython/pystate.h#L159-L165, thanks dzhulgakov for the pointer), but you'd have to do some extra work at runtime to know how to map their TID to ours so for now I'm going to stick to what I can glean from post processing alone.

As we start observing more threads it becomes more important to be principled about how we start up and shut down. (Since threads may die while the profiler is running.) #82154 had various troubles with segfaults that wound up being related to accessing Python thread pointers which were no longer alive. I've tweaked the startup and shutdown interaction with the CPython interpreter and it should be safer now.

Differential Revision: [D42336292](https://our.internmc.facebook.com/intern/diff/D42336292/)

[ghstack-poisoned]

robieta commented:
@dzhulgakov Sorry for the delay; it turns out we disable Dyno in unit tests so I had a hell of a time reproducing.

AFAICT `gil_scoped_acquire` (or CPython threading) is just kind of brittle. I added a disarm clause which seems to address the issue. (And packaged everything into a higher level guard that I also use during startup.) Let me know if it fixes the issue on your end.

dzhulgakov commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -600,6 +601,28 @@ static PyTypeObject TraceContextType = {
+
+    // `gil_scoped_acquire` is a bit fragile in on-demand mode:
+    // https://github.com/pytorch/pytorch/pull/91684#issuecomment-1413154458
+    if (!Py_IsInitialized()) {
```
Nice hack, let me try it out!

Commit:
Update on "[Profiler] Defer recording startup python events (take 2)"


This is my commandeer of https://github.com/pytorch/pytorch/pull/82154 with a couple extra fixes.

The high level idea is that when we start profiling we see python frames which are currently executing, but we don't know what system TID created them. So instead we defer the TID assignment, and then during post processing we peer into the future and use the system TID *of the next* call on that Python TID.

As an aside, it turns out that CPython does some bookkeeping (https://github.com/python/cpython/blob/ee821dcd3961efc47262322848267fe398faa4e4/Include/cpython/pystate.h#L159-L165, thanks dzhulgakov for the pointer), but you'd have to do some extra work at runtime to know how to map their TID to ours so for now I'm going to stick to what I can glean from post processing alone.

As we start observing more threads it becomes more important to be principled about how we start up and shut down. (Since threads may die while the profiler is running.) #82154 had various troubles with segfaults that wound up being related to accessing Python thread pointers which were no longer alive. I've tweaked the startup and shutdown interaction with the CPython interpreter and it should be safer now.

Differential Revision: [D42336292](https://our.internmc.facebook.com/intern/diff/D42336292/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Defer recording startup python events (take 2)"


This is my commandeer of https://github.com/pytorch/pytorch/pull/82154 with a couple extra fixes.

The high level idea is that when we start profiling we see python frames which are currently executing, but we don't know what system TID created them. So instead we defer the TID assignment, and then during post processing we peer into the future and use the system TID *of the next* call on that Python TID.

As an aside, it turns out that CPython does some bookkeeping (https://github.com/python/cpython/blob/ee821dcd3961efc47262322848267fe398faa4e4/Include/cpython/pystate.h#L159-L165, thanks dzhulgakov for the pointer), but you'd have to do some extra work at runtime to know how to map their TID to ours so for now I'm going to stick to what I can glean from post processing alone.

As we start observing more threads it becomes more important to be principled about how we start up and shut down. (Since threads may die while the profiler is running.) #82154 had various troubles with segfaults that wound up being related to accessing Python thread pointers which were no longer alive. I've tweaked the startup and shutdown interaction with the CPython interpreter and it should be safer now.

Differential Revision: [D42336292](https://our.internmc.facebook.com/intern/diff/D42336292/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Defer recording startup python events (take 2)"


This is my commandeer of https://github.com/pytorch/pytorch/pull/82154 with a couple extra fixes.

The high level idea is that when we start profiling we see python frames which are currently executing, but we don't know what system TID created them. So instead we defer the TID assignment, and then during post processing we peer into the future and use the system TID *of the next* call on that Python TID.

As an aside, it turns out that CPython does some bookkeeping (https://github.com/python/cpython/blob/ee821dcd3961efc47262322848267fe398faa4e4/Include/cpython/pystate.h#L159-L165, thanks dzhulgakov for the pointer), but you'd have to do some extra work at runtime to know how to map their TID to ours so for now I'm going to stick to what I can glean from post processing alone.

As we start observing more threads it becomes more important to be principled about how we start up and shut down. (Since threads may die while the profiler is running.) #82154 had various troubles with segfaults that wound up being related to accessing Python thread pointers which were no longer alive. I've tweaked the startup and shutdown interaction with the CPython interpreter and it should be safer now.

Differential Revision: [D42336292](https://our.internmc.facebook.com/intern/diff/D42336292/)

[ghstack-poisoned]

robieta commented:
@pytorchbot merge -f "unit test failures are unrelated"

zejun-chen commented:
Hi, @robieta 

How are you? Here is `test_source_multithreaded` UT failure issue we met when moving the cases `test_source_multithreaded` outside of the test class `TestProfiler`. We just move the entire cases but no code change is made. Could you help take a look?Here is the detailed: https://github.com/pytorch/pytorch/pull/134316#issuecomment-2337167111

Thank you.

zejun-chen commented on test/profiler/test_profiler.py:
```diff
@@ -549,6 +553,161 @@ def extract(pattern: str):
+        name_fn=lambda name, thread_spec: name
+    )
+    @parametrize("work_in_main_thread", [True, False])
+    def test_source_multithreaded(self, name, thread_spec, work_in_main_thread):
```
Hi, @robieta 
May i know if the case `test_source_multithreaded` has the internal connection with the other test cases in the same test class `TestProfiler`? We find the case `profiler/test_profiler.py::TestProfilerMultithreaded::test_source_multithreaded_basic_work_in_main_thread_True` will failed if moving the test case `test_source_multithreaded` outside of the test class `TestProfiler`.
Here is the test PR: https://github.com/pytorch/pytorch/pull/135971
Could you help take a look?
Thank you.

zejun-chen replied:
Here is the error log i got in CI:
```
2024-09-16T02:54:44.0637567Z   File "/var/lib/jenkins/workspace/test/profiler/test_profiler.py", line 228, in test_source_multithreaded
2024-09-16T02:54:44.0678730Z #22 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0679126Z #23 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0679599Z #24 _PyObject_VectorcallTstate.lto_priv.4.cold from :0
2024-09-16T02:54:44.0680087Z #25 method_vectorcall from :0
2024-09-16T02:54:44.0680454Z #26 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0680838Z #27 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0681324Z #28 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0681723Z #29 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0682132Z #30 eval_custom_code from eval_frame.c:0
2024-09-16T02:54:44.0682585Z #31 custom_eval_frame_shim from eval_frame.c:0
2024-09-16T02:54:44.0683028Z #32 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0683413Z #33 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0683837Z #34 eval_custom_code from eval_frame.c:0
2024-09-16T02:54:44.0684288Z #35 custom_eval_frame_shim from eval_frame.c:0
2024-09-16T02:54:44.0684732Z #36 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0685147Z #37 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0685635Z #38 _PyObject_VectorcallTstate.lto_priv.4.cold from :0
2024-09-16T02:54:44.0686103Z #39 method_vectorcall from :0
2024-09-16T02:54:44.0686476Z #40 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0686889Z #41 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0687274Z #42 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0687689Z #43 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0688082Z #44 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0688550Z #45 _PyObject_VectorcallTstate.lto_priv.4.cold from :0
2024-09-16T02:54:44.0689028Z #46 method_vectorcall from :0
2024-09-16T02:54:44.0689398Z #47 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0689801Z #48 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0690226Z #49 eval_custom_code from eval_frame.c:0
2024-09-16T02:54:44.0690677Z #50 custom_eval_frame_shim from eval_frame.c:0
2024-09-16T02:54:44.0691115Z #51 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0691530Z #52 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0691995Z #53 _PyObject_VectorcallTstate.lto_priv.4 from :0
2024-09-16T02:54:44.0692440Z #54 method_vectorcall from :0
2024-09-16T02:54:44.0692810Z #55 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0693287Z #56 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0693765Z #57 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0694259Z #58 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0694651Z #59 slot_tp_call from :0
2024-09-16T02:54:44.0695009Z #60 _PyObject_MakeTpCall.localalias from :0
2024-09-16T02:54:44.0695463Z #61 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0695944Z #62 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0696422Z #63 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0696893Z #64 slot_tp_call from :0
2024-09-16T02:54:44.0697242Z #65 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0697647Z #66 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0698130Z #67 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0698630Z #68 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0699007Z #69 slot_tp_call from :0
2024-09-16T02:54:44.0699377Z #70 _PyObject_MakeTpCall.localalias from :0
2024-09-16T02:54:44.0699831Z #71 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0700307Z #72 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0700793Z #73 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0701179Z #74 slot_tp_call from :0
2024-09-16T02:54:44.0701534Z #75 _PyObject_MakeTpCall.localalias from :0
2024-09-16T02:54:44.0701984Z #76 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0702465Z #77 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0702941Z #78 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0703329Z #79 slot_tp_call from :0
2024-09-16T02:54:44.0703699Z #80 _PyObject_MakeTpCall.localalias from :0
2024-09-16T02:54:44.0714578Z #81 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0715148Z #82 PyEval_EvalCode from ??:0
2024-09-16T02:54:44.0715521Z #83 run_eval_code_obj from :0
2024-09-16T02:54:44.0715867Z #84 run_mod from :0
2024-09-16T02:54:44.0716155Z #85 pyrun_file from :0
2024-09-16T02:54:44.0716549Z #86 _PyRun_SimpleFileObject.localalias from :0
2024-09-16T02:54:44.0717050Z #87 _PyRun_AnyFileObject.localalias from :0
2024-09-16T02:54:44.0717624Z #88 Py_RunMain.localalias from :0
2024-09-16T02:54:44.0718014Z #89 Py_BytesMain from ??:0
2024-09-16T02:54:44.0718364Z #90 __libc_start_main from ??:0
2024-09-16T02:54:44.0718712Z #91 _start from ??:0
```
